pkg/controller/servicemesh/webhooks/validation/controlplane_downgrade_v1_1.go

```diff
@@ -0,0 +1,100 @@
+package validation
+
+import (
+	"context"
+	"fmt"
+
+	pkgerrors "github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	maistrav1 "github.com/maistra/istio-operator/pkg/apis/maistra/v1"
+	"github.com/maistra/istio-operator/pkg/controller/common"
+)
+
+var (
+	// These are unsupported in v1.1
+	unsupportedNewResourcesV1_0 = []schema.GroupVersionKind{
+		schema.GroupVersionKind{Group: "security.istio.io", Version: "v1beta1", Kind: "AuthorizationPolicy"},
+	}
+)
+
+func (v *ControlPlaneValidator) validateDowngradeFromV1_1(smcp *maistrav1.ServiceMeshControlPlane) error {
+	var allErrors []error
+	meshNamespaces := sets.NewString(smcp.GetNamespace())
+
+	memberNamespaces := &corev1.NamespaceList{}
+	if err := v.client.List(context.TODO(), client.MatchingLabels(map[string]string{common.MemberOfKey: smcp.GetNamespace()}), memberNamespaces); err != nil {
+		return pkgerrors.Wrap(err, "error listing member namespaces")
+	}
+	for _, member := range memberNamespaces.Items {
+		meshNamespaces.Insert(member.GetName())
+		// ca.istio.io/env label exists on any member namespaces
+		if common.HasLabel(&member.ObjectMeta, "ca.istio.io/env") {
+			allErrors = append(allErrors, fmt.Errorf("ca.istio.io/env label on namespace %s is not supported in older version", member.GetName()))
+		}
+		// ca.isio.io/override label exists on any member namespaces
+		if common.HasLabel(&member.ObjectMeta, "ca.istio.io/override") {
+			allErrors = append(allErrors, fmt.Errorf("ca.istio.io/override label on namespace %s is not supported in older version", member.GetName()))
+		}
+	}
+
+	// Any VirtualService http entries use mirrorPercent attribute
+	vsgvk := schema.GroupVersionKind{Group: "networking.istio.io", Version: "v1alpha3", Kind: "VirtualService"}
+	virtualServices := &unstructured.UnstructuredList{}
+	virtualServices.SetGroupVersionKind(vsgvk)
+	// XXX: do we list all in the cluster, or list for each member namespace?
+	if err := v.client.List(context.TODO(), nil, virtualServices); err != nil {
```

Comment:
There are a couple of places where we do things like this.  Should we?

My concerns are with the number of calls to the apiserver and possibly the amount of data returned.  We query a number of deprecated resource types to ensure an upgrade won't break (we could remove that logic), and we query a number of resources to see if they're configured in an unsupported manner (e.g. http- named ports using 443).

Reply:
Should we maybe use `istio/client-go` to query the objects instead of unstructured?

Reply:
We should use caching. It's not appropriate for a webhook to produce multiple API requests. Also consider one of my other comments: should we really check for the existence of other resource types?

Reply:
We should, but I was weary of adding that as a dependency.  I'll see how much effort is involved.

Reply:
These types have been removed from istio/client-go@1.4.5, so we're stuck using unstructured.

Reply:
I see, thanks for checking!